torch/csrc/distributed/c10d/reducer.cpp

```diff
@@ -126,33 +128,34 @@ Reducer::Reducer(
         backward_stats_.end(),
         [=](std::vector<int64_t>& v) { v.resize(variable_count); });
   }
+  if (find_unused_parameters_) {
```

Comment:
Can we add a note explaining why we can avoid initializing these variables when `find_unused_parameters_` is false? We can refer to that note in each place we have such logic. Example of note: https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/engine.cpp#L100. Example of referring to a note: https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/engine.cpp#L64.

Reply:
There were two errors caused by setting `global_unused=false` when `find_unused_parameters_ = false`. I keep using `local_used_maps_` when `find_unused_parameters_ = false` to avoid the errors and now it passes all the cases. Let me know about the alternatives.
Errors were as follows:
```
..............ERROR:root:Caught exception: 
Traceback (most recent call last):
  File "/data/users/sinannasir/pytorch/torch/testing/_internal/common_distributed.py", line 207, in wrapper
    fn()
  File "/data/users/sinannasir/pytorch/torch/testing/_internal/common_distributed.py", line 76, in wrapper
    return func(*args, **kwargs)
  File "test/distributed/test_c10d.py", line 2380, in test_global_local_unused_params_grad
    run_and_verify_grad(cpu_model)
  File "test/distributed/test_c10d.py", line 2369, in run_and_verify_grad
    self.assertIsNone(task_unused_p.grad)
  File "/home/sinannasir/local/miniconda3/envs/pytorch/lib/python3.8/unittest/case.py", line 1322, in assertIsNone
    self.fail(self._formatMessage(msg, standardMsg))
  File "/home/sinannasir/local/miniconda3/envs/pytorch/lib/python3.8/unittest/case.py", line 753, in fail
    raise self.failureException(msg)
AssertionError: tensor([[0., 0.],
        [0., 0.]]) is not None
exiting process with exit code: 10
ERROR:root:Caught exception: 
Traceback (most recent call last):
  File "/data/users/sinannasir/pytorch/torch/testing/_internal/common_distributed.py", line 207, in wrapper
    fn()
  File "/data/users/sinannasir/pytorch/torch/testing/_internal/common_distributed.py", line 76, in wrapper
    return func(*args, **kwargs)
  File "test/distributed/test_c10d.py", line 2380, in test_global_local_unused_params_grad
    run_and_verify_grad(cpu_model)
  File "test/distributed/test_c10d.py", line 2369, in run_and_verify_grad
    self.assertIsNone(task_unused_p.grad)
  File "/home/sinannasir/local/miniconda3/envs/pytorch/lib/python3.8/unittest/case.py", line 1322, in assertIsNone
    self.fail(self._formatMessage(msg, standardMsg))
  File "/home/sinannasir/local/miniconda3/envs/pytorch/lib/python3.8/unittest/case.py", line 753, in fail
    raise self.failureException(msg)
AssertionError: tensor([[0., 0.],
        [0., 0.]]) is not None
exiting process with exit code: 10
Process 0 terminated with exit code 10, terminating remaining processes.
Es.....s...............................................test/distributed/test_c10d.py:1531: UserWarning: Deprecation warning: In a future PyTorch release torch.full will no longer return tensors of floating dtype by default. Instead, a bool fill_value will return a tensor of torch.bool dtype, and an integral fill_value will return a tensor of torch.long dtype. Set the optional `dtype` or `out` arguments to suppress this warning. (Triggered internally at  ../aten/src/ATen/native/TensorFactories.cpp:364.)
  tensor = torch.full([100, 100], self.rank)
test/distributed/test_c10d.py:1531: UserWarning: Deprecation warning: In a future PyTorch release torch.full will no longer return tensors of floating dtype by default. Instead, a bool fill_value will return a tensor of torch.bool dtype, and an integral fill_value will return a tensor of torch.long dtype. Set the optional `dtype` or `out` arguments to suppress this warning. (Triggered internally at  ../aten/src/ATen/native/TensorFactories.cpp:364.)
  tensor = torch.full([100, 100], self.rank)
test/distributed/test_c10d.py:1531: UserWarning: Deprecation warning: In a future PyTorch release torch.full will no longer return tensors of floating dtype by default. Instead, a bool fill_value will return a tensor of torch.bool dtype, and an integral fill_value will return a tensor of torch.long dtype. Set the optional `dtype` or `out` arguments to suppress this warning. (Triggered internally at  ../aten/src/ATen/native/TensorFactories.cpp:364.)
  tensor = torch.full([100, 100], self.rank)
test/distributed/test_c10d.py:1531: UserWarning: Deprecation warning: In a future PyTorch release torch.full will no longer return tensors of floating dtype by default. Instead, a bool fill_value will return a tensor of torch.bool dtype, and an integral fill_value will return a tensor of torch.long dtype. Set the optional `dtype` or `out` arguments to suppress this warning. (Triggered internally at  ../aten/src/ATen/native/TensorFactories.cpp:364.)
  tensor = torch.full([100, 100], self.rank)
.test/distributed/test_c10d.py:1554: UserWarning: Deprecation warning: In a future PyTorch release torch.full will no longer return tensors of floating dtype by default. Instead, a bool fill_value will return a tensor of torch.bool dtype, and an integral fill_value will return a tensor of torch.long dtype. Set the optional `dtype` or `out` arguments to suppress this warning. (Triggered internally at  ../aten/src/ATen/native/TensorFactories.cpp:364.)
  self.assertEqual(torch.full([10, 10], self.world_size), tensor)
test/distributed/test_c10d.py:1554: UserWarning: Deprecation warning: In a future PyTorch release torch.full will no longer return tensors of floating dtype by default. Instead, a bool fill_value will return a tensor of torch.bool dtype, and an integral fill_value will return a tensor of torch.long dtype. Set the optional `dtype` or `out` arguments to suppress this warning. (Triggered internally at  ../aten/src/ATen/native/TensorFactories.cpp:364.)
  self.assertEqual(torch.full([10, 10], self.world_size), tensor)
test/distributed/test_c10d.py:1554: UserWarning: Deprecation warning: In a future PyTorch release torch.full will no longer return tensors of floating dtype by default. Instead, a bool fill_value will return a tensor of torch.bool dtype, and an integral fill_value will return a tensor of torch.long dtype. Set the optional `dtype` or `out` arguments to suppress this warning. (Triggered internally at  ../aten/src/ATen/native/TensorFactories.cpp:364.)
  self.assertEqual(torch.full([10, 10], self.world_size), tensor)
test/distributed/test_c10d.py:1554: UserWarning: Deprecation warning: In a future PyTorch release torch.full will no longer return tensors of floating dtype by default. Instead, a bool fill_value will return a tensor of torch.bool dtype, and an integral fill_value will return a tensor of torch.long dtype. Set the optional `dtype` or `out` arguments to suppress this warning. (Triggered internally at  ../aten/src/ATen/native/TensorFactories.cpp:364.)
  self.assertEqual(torch.full([10, 10], self.world_size), tensor)
.....s...............F...............
======================================================================
ERROR: test_global_local_unused_params_grad (__main__.DistributedDataParallelTest)
By simulating a multi-task training, this test is to make sure:
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/data/users/sinannasir/pytorch/torch/testing/_internal/common_distributed.py", line 204, in wrapper
    self._join_processes(fn)
  File "/data/users/sinannasir/pytorch/torch/testing/_internal/common_distributed.py", line 306, in _join_processes
    self._check_return_codes(elapsed_time)
  File "/data/users/sinannasir/pytorch/torch/testing/_internal/common_distributed.py", line 339, in _check_return_codes
    raise RuntimeError(error)
RuntimeError: Processes 0 1 exited with error code 10

======================================================================
FAIL: test_forward_backward_unused_parameters (__main__.ReducerTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test/distributed/test_c10d.py", line 3049, in test_forward_backward_unused_parameters
    self.assertEqual(None, model.fc3.weight.grad)
  File "/data/users/sinannasir/pytorch/torch/testing/_internal/common_utils.py", line 1124, in assertEqual
    super().assertEqual(x, y, msg=msg)
AssertionError: None != tensor([[0., 0., 0., 0.],
        [0., 0.[58 chars]0.]])

----------------------------------------------------------------------
Ran 107 tests in 175.165s

FAILED (failures=1, errors=1, skipped=3)
```

Reply:
hmm, I checked `test_global_local_unused_params_grad`, and looks like it is setting `find_unused_parameters` to true. So in this case `global_unused` should not be directly set to `false` no?

Reply:
For `test_forward_backward_unused_parameters`, if you look at how the reducer is created, it does not pass `find_unused_parameters` properly, and relies on `prepare_for_backward` to detect that using the forward `output`. So you will need to modify the test accordingly.

Reply:
> For `test_forward_backward_unused_parameters`, if you look at how the reducer is created, it does not pass `find_unused_parameters` properly, and relies on `prepare_for_backward` to detect that using the forward `output`. So you will need to modify the test accordingly.

I see your point. For those cases, it was setting `find_unused_parameters_` to False by default and not passing the actual value. I will revert reducer.cpp back and keep avoiding initializing `local_used_parameters_`, and look into the test cases. Thanks,

Reply:
> hmm, I checked `test_global_local_unused_params_grad`, and looks like it is setting `find_unused_parameters` to true. So in this case `global_unused` should not be directly set to `false` no?

I was declaring global_unused twice when `find_unused_parameters=true`. I fixed that and now all test cases pass without any modification on `test_forward_backward_unused_parameters`. I still modified `test_forward_backward_unused_parameters` and pass `find_unused_parameters` properly for correctness.

Reply:
> Can we add a note explaining why we can avoid initializing these variables when `find_unused_parameters_` is false? We can refer to that note in each place we have such logic. Example of note: https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/engine.cpp#L100. Example of referring to a note: https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/engine.cpp#L64.

Working on this feedback.